packages/core/core/src/public/Asset.js

```diff
@@ -0,0 +1,167 @@
+// @flow strict-local
+// flowlint unsafe-getters-setters:off
+
+import type {Readable} from 'stream';
+
+import type {
+  Asset as IAsset,
+  AST,
+  Config,
+  Dependency,
+  DependencyOptions,
+  Environment,
+  File,
+  FilePath,
+  Meta,
+  MutableAsset as IMutableAsset,
+  PackageJSON,
+  SourceMap,
+  Stats
+} from '@parcel/types';
+
+import type InternalAsset from '../Asset';
+
+import nullthrows from 'nullthrows';
+
+const _assetToInternalAsset: WeakMap<Asset, InternalAsset> = new WeakMap();
+
+export function assetToInternalAsset(asset: Asset): InternalAsset {
+  return nullthrows(_assetToInternalAsset.get(asset));
+}
+
+export class Asset implements IAsset {
+  #asset; // InternalAsset
+
+  constructor(asset: InternalAsset) {
+    this.#asset = asset;
+    _assetToInternalAsset.set(this, asset);
+  }
+
+  get id(): string {
+    return this.#asset.id;
+  }
+
+  get ast(): ?AST {
+    return this.#asset.ast;
+  }
+
+  get type(): string {
+    return this.#asset.type;
+  }
+
+  get env(): Environment {
+    return this.#asset.env;
+  }
+
+  get filePath(): FilePath {
+    return this.#asset.filePath;
+  }
+
+  get meta(): Meta {
+    return this.#asset.meta;
+  }
+
+  get outputHash(): string {
```

Comment:
Should this be internal?

Reply:
It's used here:

https://github.com/parcel-bundler/parcel/blob/b96d50c4311ed7bec2859150233991648068044d/packages/namers/default/src/DefaultNamer.js#L83

Maybe we separate Asset into `TransformerAsset` and `CompletedAsset` (or similar), where only the latter has `outputHash`?